Use static const for default log level so Swift can strip strings

ffried opened the pull request:
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/CocoaLumberjack/)
* [x] I have searched for a similar pull request in the [project](https://github.com/CocoaLumberjack/CocoaLumberjack/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: #774 

### Pull Request Description

As described in #774, the previous `defaultDebugLevel` was a normal global variable, that could change at any time. Therefor the compiler did not strip strings that were logged with an inaccessible level.
This changes this by adding `@inlinable` to the Swift log functions (so the compiler is allowed to inline the code) and using a `static const` definition for the default log level. This definition now cannot be changed in Swift code, but needs to be defined either via the `GCC_PREPROCESSOR_DEFINITIONS` build setting, or using a `#define` in a bridging header of a Swift target.

rivera-ernesto commented:
Looking good. Still WIP? Feel free to make multiple PR's.

ffried commented:
Mainly still WIP since I need to update the docs (which I could do in another PR however).

But I think we need to be careful on how we integrate this. This will break existing Swift code and remove the possibility to change the default log level at runtime in Swift.
I don't know if this is something that is needed. If so, we could add another layer below the `static const` (which of course then wouldn't strip strings).
So from a release perspective, this should go at least into a new minor version. If we want to stick strictly to semantic versioning, it should even go into a major release.

ffried commented:
To elaborate a bit on how to re-add changing the log level at runtime:
I'd re-add the `defaultDebugLevel` (though probably rename it, since it's not only for _debug_) and add the bitmask-check *after* the bitmask-check of the `static const`. Meaning that if the `static const` is disabling logging for the level of a log message, all bets are off and it's not going to be logged (and in release builds also stripped from the binary). If the `static const` allows the logging of a given message, however, it'll then be checked against the global Swift `var`. And only if both checks succeed, it'll be passed to `DDLog`.
Of course, this also introduces more complexity (e.g. questions like "why is my verbose log message not logged, I've set `defaultDebugLevel` to `.all`?" which could happen because `DD_LOG_LEVEL` is set to `DDLogLevelWarning`). On the other hand if we also make the default `.all` for `defaultDebugLevel`, this should only happen if both `defaultDebugLevel` *and* `DD_LOG_LEVEL` have already been changed. And at that point I'd expect the developer to know that.

ffried commented:
So, to summarize: this one right here is ready to be merged (minus the docs which I'd add in another PR).
If we want to re-add changing the log level in Swift at runtime, I'd also create another PR for this.
This would then also make it possible to ship this in a minor release (since it does not entirely remove a feature).

sushichop commented:
@ffried 

Thanks for explaining in detail😁

> But I think we need to be careful on how we integrate this. This will break existing Swift code and remove the possibility to change the default log level at runtime in Swift.

I agree with you. I concerned about this when I looked at this PR.

> If we want to re-add changing the log level in Swift at runtime, I'd also create another PR for this.

Great!! IMO, we (and many developers) need to be able to change log level in Swift at runtime.
I think this PR can be merged if you create another PR for it.

Then, which should we merge this PR right now or wait for another PR? 
(Can you let us know which one you prefer?)

Thanks in advance.

ffried commented:
@sushichop Thanks for confirming my concerns. Wasn't sure if I'm the only one having them. 🙂 


> Then, which should we merge this PR right now or wait for another PR?
(Can you let us know which one you prefer?)

This depends a bit on how close the next release is. I think I should be able to create the follow-up PR within the next week or so. So if there's another release planned until then, better wait with merging this PR. Otherwise this may as well be merged.

ffried commented:
On second thought, maybe it's better if I extend this PR (instead of creating another one). Otherwise the follow-up PR will just partially revert what I deleted in this PR. Especially since I basically need some of the now deleted Swift code (even if I rename the global var, I need to deprecate the old for a smooth transition).

sushichop commented:
@ffried 
OK. It seems that we just have better wait for this PR to be extended.

@rivera-ernesto @bpoplauschi 
What do think about this?

ffried commented:
@sushichop Done from my side. There is now a `dynamicLogLevel` with which a message can be "filtered" at runtime. It is, however, not used for setting the level of messages (like the `defaultDebugLevel` did before). The level of messages will always be set from `DDDefaultLogLevel`.

---
No idea why the codebeat check does not finish...

sushichop commented:
@ffried 
Thanks. LGTM😁

> No idea why the codebeat check does not finish...

I also have no idea...

@bpoplauschi @rivera-ernesto 
Do you know something about codebeat or codacy? 
And can you review this PR? Now, I think we are able to merge it.

rivera-ernesto commented:
> Do you know something about codebeat or codacy?

I added both. They just check coding style, common bugs, etc.
They just give us and contributors a second/third/fourth opinion on the PR.

> And can you review this PR? Now, I think we are able to merge it.

I tend to merge often. We have many users and if something's missing or wrong a new PR will come right after.

ffried commented:
Thanks! I'll create another PR with the docs update ASAP.

sushichop commented:
@rivera-ernesto 
All right. Thanks😁

bpoplauschi commented:
Very good contribution @ffried 
The only open question is: do we need a major release (i.e. 4.0.0) for this feature? I think it changes the library's API, but I'm not sure.

ffried commented:
@bpoplauschi now we don’t need a major release anymore. We would have with my first approach where I completely removed the possibility to change the log level in Swift at runtime. But since I added it back with a new variable name and properly deprecated the old one, this can also be released in a minor release now.
It still slightly changes the API now, but in a backwards compatible manner.

bpoplauschi commented:
@ffried perfect, I missed the adding back part :)
Then probably a minor release (3.5.0)
